### ISSUE TITLE
New editor group split sizing feature misbehaving

### ISSUE THREAD

**Issue by sebastianzander** [microsoft/vscode#187431]
Type: <b>Bug</b>

Just like in the demonstration of Release Notes: 1.80.0, create two initial editor groups, manually move the split in order to create a custom layout. Click in the left editor group and choose "Split editor right". The left editor group is further split in the middle, leaving the initial split where it is. Now close the middle (latest) editor split group again. This will **[actual]** merge the middle and the right editor groups, although it should **[expected]** merge the middle and the left editor group, since the left editor group was split previously and gave off the space.

This disturbs any custom layout immensely.

VS Code version: Code 1.80.0 (660393deaaa6d1996740ff4880f1bad43768c814, 2023-07-04T15:06:02.407Z)
OS version: Windows_NT x64 10.0.19045
Modes:
Remote OS version: Linux x64 5.15.0-76-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz (8 x 2304)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.76GB (5.48GB free)|
|Process Argv|--folder-uri vscode-remote://ssh-remote%2Bhco-s4rj0501.local.hacon.de/opt/sda/tps --crash-reporter-id 96988c55-f48b-4883-ab81-cd771e3153f5|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: hco-s4rj0501.local.hacon.de|
|OS|Linux x64 5.15.0-76-generic|
|CPUs|Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz (16 x 2900)|
|Memory (System)|78.26GB (59.38GB free)|
|VM|0%|
</details><details><summary>Extensions (54)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.3.1
vscode-custom-css|be5|7.2.1
vsc-material-theme|Equ|33.8.0
vsc-material-theme-icons|equ|2.8.0
vscode-drawio|hed|1.6.6
vscode-edge-devtools|ms-|2.1.2
jupyter-keymap|ms-|1.1.2
remote-containers|ms-|0.299.0
remote-ssh|ms-|0.102.0
remote-wsl|ms-|0.80.2
remote-explorer|ms-|0.4.1
remote-server|ms-|1.2.1
material-icon-theme|PKi|4.28.0
restore-editors|amo|1.0.1
clipboard-history|Anj|1.0.7
working-sets|ber|2.0.0
multi-cursor-case-preserve|Car|1.0.5
flip|Chr|0.0.4
doxdocgen|csc|1.4.0
xslt-xpath|del|1.5.13
gitlens|eam|2023.7.705
swap|fau|0.0.5
autoscrolldown|fox|1.2.1
duplicate|gee|1.0.2
gitlab-workflow|Git|3.68.0
selected-lines-count|gur|1.4.0
cmake-language-support-vscode|jos|0.0.9
indent-to-bracket|LAK|1.2.2
vscode-clangd|llv|0.1.24
tasknotify|lwl|0.0.2
git-graph|mhu|1.30.0
prettify-json|moh|0.0.3
vscode-dotnet-runtime|ms-|1.6.0
isort|ms-|2023.10.0
python|ms-|2023.12.0
vscode-pylance|ms-|2023.7.10
jupyter|ms-|2023.6.1001861915
jupyter-keymap|ms-|1.1.2
jupyter-renderers|ms-|1.0.17
vscode-jupyter-cell-tags|ms-|0.1.8
vscode-jupyter-slideshow|ms-|0.1.5
cpptools|ms-|1.16.3
cpptools-extension-pack|ms-|1.3.0
hexeditor|ms-|1.9.11
clang-tidy|not|0.5.1
jinjahtml|sam|0.20.0
vscode-fileutils|sle|3.10.2
vscode-deepl|soe|1.0.6
code-spell-checker|str|2.20.5
sort-lines|Tyr|1.10.2
vscode-lldb|vad|1.9.2
expand-selection-to-scope|vit|0.2.0
change-case|wma|1.0.0
save-as-root|yy0|1.6.0

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30410667
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
282f8724:30602487
pyind779:30671433
f6dab269:30613381
pythonsymbol12:30671437
a9j8j154:30646983
showlangstatbar:30737416
vsctsb:30748421
pythonms35:30701012
03d35959:30757346
pythonfmttext:30731395
pythoncmv:30756943
fixshowwlkth:30771522
hideindicator:30785051
pythongtdpath:30769146
bgfeh915:30780428
pythonnosm12tcf:30779713
pythonidxpt:30784022
pythonnoceb:30776495
h7j2d465:30772216
cmakesidepanelv2:30779593

```

</details>

**Comment by sebastianzander** [comment]
Not knowing how editor group splitting is implemented internally, I have some assumptions and could see a preferred solution. I used tuple formatting to visualize the width's of editor groups in order like `(left, middle, right)`:

- when you manually move the splitter you set a width (absolute or relative, possibly) for both affected editor groups, since you don't know which editor group you want to have a certain width (could be both of course):  `(n%, m%)`
- when you further split one of the editor groups, it should again get an automatic width just like the newly created editor group, e.g. if you split the left of the initial two editor groups: `(auto, auto, m%)`
- when you close any of the two editor groups with auto width, they should merge and leave the third editor group that still has some width setting intact: `(auto, m%)` or again calculated to `(100% - m%, m%)`

Not sure if this is feasible and compatible with the current implementation, though.

**Comment by ackvf** [comment]
1. have two tab groups with custom layout (move the boundary)
2. split **left group** to the right 
    - _(expected)_ observe that the left group it split in two, keeping the original boundary in place
3. close the new group
   - the **right** group reclaims the space and not the left as would be expected

![vscode-split-group-right](https://github.com/microsoft/vscode/assets/2313018/432e059e-8f06-4db1-872f-6525d918dd28)


1. ditto
2. split **left group** to the left
    - inconsistently with the previous example, **all editor groups resize** proportionally
3. close the new group
   - the **rightmost** group reclaims **all the space** and not the "original left" as would be expected
   - the left group remains resized

![vscode-split-group-left](https://github.com/microsoft/vscode/assets/2313018/e319de12-aac4-4010-8f0e-c0eb3166b684)

**Comment by ackvf** [comment]
This same inconsistent and unexpected behavior can be observed with vertically split editors as well.

**Comment by ackvf** [comment]
This same inconsistent and unexpected behavior can be observed with vertically split editors as well.

**Comment by bpasero** [comment]
Recording to demonstrate that the initial layout cannot be restored when closing the second editor group:

![Recording 2023-07-21 at 14 40 49](https://github.com/microsoft/vscode/assets/900690/b3e8f196-3e37-40fa-9bce-8d6556600df8)

From the editor grid, we call `removeView` with the sizing option `auto` that we added:

https://github.com/microsoft/vscode/blob/110e2f633cb23724166269c9102e7969df271dc7/src/vs/workbench/browser/parts/editor/editorPart.ts#L720

Moving to grid boss for consideration 🦑

**Comment by joaomoreno** [comment]
Would this be better?


https://github.com/microsoft/vscode/assets/22350/fbfec25c-39bc-41c6-a572-72cd81cb950d

**Comment by sebastianzander** [comment]
> Would this be better?
This looks perfect. It seemed intuitive to me the whole time 👍

**Comment by ackvf** [comment]
@joaomoreno Amazing!!! these below are my opinions, not a directive, forgive my crude form

0:05 creating 3 new split editors consecutively should create 3 equal editors ✔
<0:12 adjusting boundaries by hand should lock them in place ✔
0:13 closed editor should allow the **closest top or left** editor to reclaim the empty place ✔
0:16 splitting an editor into two should create two editors in equal size, not adjusting size of adjacent editors IF the boundary was manually adjusted (locked) ❌
0:21 closed editor should allow the **closest top or left** editor to reclaim the empty place ✔
0:22 splitting an editor into two should create two editors in equal size, not adjusting size of adjacent editors IF the boundary was manually adjusted (locked) ✔
0:25 closed leftmost/topmost editor should allow the closest right/bottom editor to reclaim the empty place ✔
0:29 closed editor should allow the **closest top or left** editor to reclaim the empty place ✔
0:30 splitting an editor into two should create two editors in equal size, not adjusting size of adjacent editors IF the boundary was manually adjusted (locked) ✔
0:35 the fourth editor adjusting widths of all other editors was a bit unexpected at first, but then I realised that none of the editors had manually adjusted (locked) boundary, so that was fine ✔, however, just as a note, if in this case, some of the boundaries were manually adjusted (locked), only the remaining editors would yield space to the new editor ⭕

**Comment by joaomoreno** [comment]
> 0:16 splitting an editor into two should create two editors in equal size, not adjusting size of adjacent editors IF the boundary was manually adjusted (locked) ❌

@bpasero I thought the implementation was correct, but now that I read this user's description, I notice he's right as well. The problem is that the setting description doesn't reflect the behavior we have implemented in the grid. The setting says:

> Splits all the editor groups to equal parts unless a part has been changed in size.

This is actually misaligned with the implementation. A better description would be:

> Splits the active editor group to equal parts, unless all editor groups are already in equal parts. In that case, splits all the editor groups to equal parts.

Unless you think the implementation should align with the setting description instead, and track user manual changes?

**Comment by ackvf** [comment]
So, what is the ultimate outcome?
1) fixing implementation to match description
   > Splits all the editor groups to equal parts unless a part has been changed in size.

or

2) keeping current implementation and updating description
   > Splits the active editor group to equal parts, unless all editor groups are already in equal parts. In that case, splits all the editor groups to equal parts.


Although, after reading it for the tenth time and trying to understand the subtleties, I find the two increasingly more similar.
IIUC, the only difference is:

1) 
   - If there are four editors `|_1_|_2_|_3_|__4__|`, with the boundary between 3|4 having been adjusted.
A new editor being created in 1 or 2 or 3 would result in `|__|__|__|__|__4__|` - in other words: |1,2,3+New editor| resizing to equal parts.
Basically what this does is split all editors to equal parts until an editor with adjusted boundary is reached.
  - If 1|2 was adjusted and new editor created in |1|, only this editor would give space to the new editor and split in half.
2) 
   - In the same situation as above `|_1_|_2_|_3_|__4__|`, a new editor being created in editor 1 would result in only editor 1 being adjusted as it gives half of its space to the New editor resulting in `|_|_|_2_|_3_|__4__|`, regardless of any manual adjustments.
   - If at that moment, all editors were equal size `|_1_|_2_|_3_|_4_|`, it would instead adjust ALL editors`|__|__|__|__|__|` to equal parts.

Is my understanding correct? In that case I think I would rather prefer option 1) keep current description and adjust implementation.

**Comment by bpasero** [comment]
Just fyi, the fix from Joao is in our insiders builds for people to play with: https://code.visualstudio.com/insiders/

**Comment by ackvf** [comment]
Thank you, so it's not my preferred option 1), but I love it anyway. Awesome job guys! This is so much better now.
